### ISSUE TITLE
Periodical BC and the GetVertex function 

### ISSUE THREAD

**Issue by Tomcat4891** [mfem/mfem#4299]
Hello.

I have a problem with getting tetrahedron coordinates from a periodic mesh. 

I have created a tetrahedral mesh for a 3d cube.
With the help of the function Mesh::MakePeriodic set periodicity for the planes XZ,YZ (four planes are pairwise periodic).

Question:
Why do I get wrong coordinates for periodic nodes if I use the following code ?
```
mesh->GetElement(i)->GetVertices(tetra);
for(int j=0;j<4;j++)
{
vertex=mesh->GetVertex(tetra[j]);
}
```

The coordinates will be correct if you use the code below:

```
ParGridFunction Answer // this was the answer to the solution of an elasticity problem
ParGridFunction nodes(Answer);
mesh->GetNodes(nodes);
mesh->GetElement(i)->GetVertices(tetra);
for(int j=0;j<4;j++)
{
vertex2[0]=nodes[3*tetra[j]];
vertex2[1]=nodes[3*tetra[j]+1];
vertex2[2]=nodes[3*tetra[j]+2];
}
```
_vertex[0], vertex[1], vertex[2]_ will be incorrect, whereas 
_vertex2[0], vertex2[1], vertex2[3]_ are defined correctly. 

Only one of the components of coordinates of periodic grid nodes is calculated incorrectly.

This is strange to me, as the documentation doesn't explicitly say this. Perhaps I'm missing something.

The problem only occur for a periodic grid, after using the MakePeriodic function.

Thanks.

Best regards,
Andrew

**Comment by najlkin** [comment]
Hi, the thing is that periodic meshes rely on discontinuous L2 maps (nodal function), which do the trick. The vertices are wrapped around while the map continues further, so you should not use vertices any longer and retrieve the coordinates only from the nodal function 👍 . You are right this is not mentioned in documentation anyhow, maybe we should fix it 🤔 , but if you look into the code of `MakePeriodic`, it is the line `periodic_mesh.SetCurvature(nodal_order, true);` which switches to the nodal function.

**Comment by najlkin** [comment]
Ok, I made a PR for it #4302 . Please check if it makes sense to you :wink:.

**Comment by najlkin** [comment]
Btw, you may also use `Mesh::GetNode()` , which is safe with and without a nodal function :wink:

**Comment by Tomcat4891** [comment]
I tried `Mesh::GetNode` and I have a different result from 

```
ParGridFunction Answer // this was the answer to the solution of an elasticity problem
ParGridFunction nodes(x);
mesh->GetNodes(nodes);
mesh->GetElement(i)->GetVertices(tetra);
for(int j=0;j<4;j++)
{
vertex2[0]=nodes[3*tetra[j]];
vertex2[1]=nodes[3*tetra[j]+1];
vertex2[2]=nodes[3*tetra[j]+2];
}
```

If you remove MakePeriodic, the results are the same. 

It's weird. Why is that? What is the correct result?

**Comment by najlkin** [comment]
This is not right, this would work if you have H1 nodal function, but `MakePeriodic()` uses discontinuous L2 function, which does not have DOFs at the vertices. If you prefer to use the nodal grid function directly instead of `GetNode`, you should interpolate it as any other grid function through `GetValue()` or you may use the method `GetNodalValues()` to get values at the vertices :wink:.

**Comment by Tomcat4891** [comment]
Thank you, for your answers. Unfortunately, I'm getting confused. I have a simple problem. Suppose I have calculated the solution of the elasticity problem on a periodic grid. I need to take a tetrahedron, get the coordinates of four points and the components of the solution for these four points (displacement vector).

Next I want to interpolate the solution inside the tetrahedron using barycentric coordinates. Seems simple, but because of the periodicity and what you said above, I am confused. 

I want to do my own fast procedure. But I can't get the trivial values of the vertices of the tetrahedra and the solution at those vertices, I don't want to use interpolation as it is time consuming for large number of points.

**Comment by najlkin** [comment]
What makes MFEM interpolation that slow? You may interpolate the solution using the builtin methods, which will work even with L2 nodal functions. However, if you insist you want to build it by yourself, I should correct myself, `Mesh::SetCurvature()` actually uses L2 elements with GaussLobatto basis, so the DOFs are actually at the vertices. You may access them through `GridFunction::GetElementDofValues()` without interpolation.

**Comment by Tomcat4891** [comment]
In general, the interpolation in the MFEM is quite fast. I need to interpolate the solution from a finite element mesh to a uniform mesh. In this case it is possible to take into account this particular feature of the problem and significantly speed up the calculations (about 50 - 100 times).

**Comment by najlkin** [comment]
I see, so does it work with `GridFunction::GetElementDofValues()`?

**Comment by Tomcat4891** [comment]
Yes, it works ! Thank you very much for your help.

**Comment by najlkin** [comment]
Perfect! :wink: I am closing this in favor of #4302 (and already merged mfem/web#264 ).
